torch/_inductor/ir.py

```diff
@@ -2084,11 +2084,16 @@ def create(cls, x, dim, start, end, step=1):
         start = cls.handle_negative_index(start, new_size[dim])
         end = cls.handle_negative_index(end, new_size[dim])
 
-        end = sizevars.evaluate_min(end, new_size[dim])
-        start = sizevars.evaluate_min(start, end)
```

Comment:
Isn't this necessary for correctness? I think the correct fix would be to use `sympy.Min` instead of `evaluate_min`.

Reply:
Thanks for the heads-up! Yeah, these lines seem relevant for unbacked SymInts, too (although not sure how well will `sympy.Min`'s application work downstream). @andrewlee302 please take a look, thanks!

Reply:
`Sympy.Min` replacement doesn't work well here. Briefly, codegen fails. Detailed logs below:
```
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL] Error in codegen for ComputedBuffer(name='buf3', layout=FixedLayout('cuda', torch.float32, size=[], stride=[]), data=Reduction(
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL]   'cuda',
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL]   torch.float32,
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL]   def inner_fn(index, rindex):
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL]       r0 = rindex
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL]       tmp0 = ops.load(arg1_1, r0 + Min(0, i9))
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL]       return tmp0
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL]   ,
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL]   ranges=[],
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL]   reduction_ranges=[-Min(0, i9) + Min(32, i9)],
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL]   reduction_type=sum,
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL]   origin_node=sum_1,
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL]   origins={sum_1}
[2023-11-12 13:03:42,567] [0/0] torch._inductor.scheduler: [CRITICAL] ))
```

I think here are two possible implications that we need to figure it out:
1. This short circuiting works well for backed symint. Is it necessary to need it and what's the benefits for the quick return? As @ezyang introduced it [here](https://github.com/pytorch/torchdynamo/commit/2788fccfc6142fe2bf1b128d9f1001402d90886e#diff-194018e9855f41edb48ecad770b7b3aa6439f6237f640d273bf1e3aa0e8377ba) and he may answer it.
2. Sympy on unbacked symint makes the codegen not work. Is it expected from the workflow perspective or sympy should be well supported for resolving this fix? 

I think both questions can be answered and this fix should be reasonable for now. We can move further later. @adnanaziz  @ezyang @peterbell10 WDTY?

Reply:
@andrewlee302 I think, the short-circuiting part is not very important: we can either keep it behind newly-added `if` statement checking that there are no unbacked SymInts in the `start` and `end`. Or, as @peterbell10 suggested, we can also drop it.

Regarding your error, I think, we may want to include unbacked SymInts (i.e., "i") in the tuple [here](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/triton.py#L1145-L1146) to circumvent them same as we currently circumvent backed SymInts (i.e., "s"). @ezyang does this sound sound to you?

But then we hit another error during compilation of the generated Triton code:

```
...
  File "/data/users/aakhundov/pytorch/torch/_inductor/codecache.py", line 887, in _run_from_cache
    return compiled_graph.compiled_artifact(inputs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/torchinductor_aakhundov/r4/cr4vabmhar475qxxdzxs5bd5tgi5plnogjqe2ngloi4b36p4cxy5.py", line 89, in call
    triton_red_fused_sum_0_rnumel = ((-1)*(Min(0, i9))) + (Min(32, i9))
                                           ^^^
NameError: name 'Min' is not defined
```

Seems that `sympy.Min` is not properly rendered in the Python wrapper codegen. @andrewlee302 I'd try adding min to the `PythonPrinter` [here](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/common.py#L340).

Reply:
So, the first thing to figure out, is what we actually want the code to be generated in this case to actually have a min/max. I find that often I reach for truly symbolic min/max when actually I can infer that a guard should go one way or another at a higher level. This is the preferred resolution and sidesteps the codegen problems here.

Looking at the tests, it seems to me that it should be statically obvious that you are within start/end. So it seems to me that we should be able to do some `evaluate_static_expr` business to get True/False without stubbing our toe on guard on unbacked symint. For examples of me having done similar refinements in the past, check https://github.com/pytorch/pytorch/pull/110979 https://github.com/pytorch/pytorch/pull/108685

Reply:
```python
    triton_red_fused_sum_0_rnumel = ((-1)*(Min(0, i9))) + (Min(32, i9))
```

`Min(0, i9)` probably shouldn't be there, sympy should be able to simplify it away to 0 if the symbol `i9` was created with the `nonnegative=True` assumption.

`Min(32, i9)` on the other hand is more tricky, we should be able to do some range analysis to bound `i9` but in the completely general case I think we *do* want to codegen a `min` here.

Reply:
I don't buy it. If the slice is in bounds, the result here should always be i9. Because the slices are coming from a torch.split, we know they are guaranteed to be in bounds after we've done error checking. The error checking itself must be done at runtime, but by the time we get to the kernel we know they must have passed.

(BTW: we aren't actually emitting the deferred runtime asserts, this is still on my list of things to do.)

Reply:
@ezyang to clarify, the codegened `Min` is not incorrect in this case (modulo the required extension of the `PythonPrinter`), just redundant? As I understand, `torch.split` does range checks and, due to those, in the generated code we can be sure that the split indices are in-range. But the deferred runtime asserts are not emitted currently. So would it make sense to keep the redundant `Min`s codegened for now and land this PR to unblock; and then remove redundant `Min`s in a follow-up, when we start emitting runtime asserts?

Reply:
The codegenned min is "correct", but it is unnecessary.

You are welcome to unblock this by putting in min codegen, if you can get it to work. But if it seems difficult to do (which it does seem like it is causing some trouble), maybe better to deal with the root cause rather than bashing through it.

Reply:
Thanks for the great input. I'm very new to pytorch and trying to interpret the discussion. Please correct me if I was wrong.
1. Dynamo does the error check for `split_with_sizes` op ([code pointer](https://github.com/pytorch/pytorch/blob/9b0f2f8d94b7485c2c2a4e3d1294a3589b3466b0/torch/_decomp/decompositions.py#L1221-L1248) before triton kernel codegen. I feed `torch.split(x, [35,-3])` and `torch.split(x, [32])` to the test and it proves that the error check happens in step 1. (logged as `[INFO] Step 1: torchdynamo start tracing`). 
2. Due to #1 that assert the range should be valid, the downstream calculation is redundant. E.g. triton code `triton_red_fused_sum_0_rnumel = ((-1)*(Min(0, i9))) + (Min(32, i9))` can be theoretically equivalent to `triton_red_fused_sum_0_rnumel = 32`.
3. But triton python code is problematic due to the lack of `from sympy import Min`. To make it runnable, we need to refine the triton codegen component. My quick question here: how difficult it is?
4. The [sizevars.evaluate_min](https://github.com/pytorch/pytorch/blob/9b0f2f8d94b7485c2c2a4e3d1294a3589b3466b0/torch/_inductor/ir.py#L2087-L2088) within `SliceView.create` happens after error handling for `torch.split` op. If all ops depending on `SliceView` can safeguard the assertion, looks like we can remove `sizevars.evaluate_min` sentences. Can we?
5. The short-circuiting is not a necessary for functionality or performance. We should get rid of it to simplify the code.

@ezyang @adnanaziz @peterbell10 WDYT?

Reply:
@ezyang Thanks for being flexible :) I think, the only thing that is missing to make the codegen-ed `sympy.Min` work is a small extension of the [`PythonPrinter`](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/common.py#L340) to add `_print_Min` (similar to the existing [`_print_Max`](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/common.py#L371-L373)). This makes @andrewlee302's unit test pass for me even when `sizevars.evaluate_min` calls are outside the new `if` statement.

@andrewlee302 IIUC, 1. and 2. sound about right. For 3., it's easy: please see the above paragraph / give it a try in the PR. For 4., I don't think we can remove `sizevars.evaluate_min` completely, as (IIUC) the runtime assertions are not yet generated from the `torch._check*`s; you can also see lots of unit tests failing now as you've removed the code in question. For 5., this is what @peterbell10 has suggested, yeah.

Reply:
yes, just update python printer